assets/__tests__/rendered_test.js

```diff
@@ -0,0 +1,191 @@
+import { Rendered } from '../js/phoenix_live_view'
+
+describe('Rendered', () => {
+  describe('mergeDiff', () => {
+    it('recursively merges two diffs', () => {
+      expect(Rendered.mergeDiff(simpleDiff1, simpleDiff2)).toEqual(
+        simpleDiffResult
+      )
+      expect(Rendered.mergeDiff(deepDiff1, deepDiff2)).toEqual(deepDiffResult)
+    })
+    it('returns the latter diff if it contains a `static` key', () => {
+      const diff1 = { 0: ['a'], 1: ['b'] }
+      const diff2 = { 0: ['c'], static: 'c' }
+      expect(Rendered.mergeDiff(diff1, diff2)).toEqual(diff2)
+    })
+  })
+
+  describe('isNewFingerprint', () => {
+    it('returns true if `diff.static` is truthy', () => {
+      const diff = { static: ['<h2>'] }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(true)
+    })
+    it('returns false if `diff.static` is falsy', () => {
+      const diff = { static: undefined }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(false)
+    })
+    it('returns false if `diff` is undefined', () => {
+      expect(Rendered.isNewFingerprint()).toEqual(false)
+    })
+  })
+
+  describe('toString', () => {
+    it('stringifies a diff', () => {
+      expect(Rendered.toString(simpleDiffResult).trim()).toEqual(
+        `
+<div class="thermostat">
+  <div class="bar cooling">
+    <a href="#" phx-click="toggle-mode">cooling</a>
+    <span>07:15:04 PM</span>
+  </div>
+  <div class="controls">
+    <span class="reading">72</span>
+    <button phx-click="dec" class="minus">-</button>
+    <button phx-click="inc" class="plus">+</button>
+    <span class="weather">
+      <div conn id="phx-dafeTrsp/m0=:DemoWeb.WeatherLive"
+     data-phx-parent-id="phx-dafeTrsp/m0="
+     data-phx-view="DemoWeb.WeatherLive"
+     data-phx-session="STUBBED"></div>
+<div class="phx-loader"></div>
+
+    </span>
+  </div>
+</div>
+      `.trim()
+      )
+    })
+  })
+
+  describe('toOutputBuffer', () => {})
```

Comment:
Haven't added anything for these three yet. They seem pretty deep down into implementation details. Maybe they don't need testing since they're not a part of the public API? Thoughts?

Reply:
This is the HTML string that is passed to morphdom.  So probably a good idea to have 1 test.  There may be an alternate implementation we "may" think about, so no need to overdue this one!